### PR TITLE
Add bombchu bag item and bombchu drops

### PR DESCRIPTION
**Pull request by mracsys**
Adds a bombchu bag model that is used for every bombchu item until one is picked up. The text for the bag informs the player that they can cut tall grass to find more bombchu drops:

image

Bombchu drops are implemented the same as @Roman971's branch (with permission). If a random drop would have been a bomb refill, it has a chance to be a bombchu refill instead. Which one is chosen is influenced by current bomb and bombchu ammo count, prioritizing refills on the lower count up to 15 of each, after which the chosen drop is completely random.

image

From discord discussion on the feature:

Complete (but still open for debate)
- Chus in logic as a setting disappears
- Anything that relies on explosives (aside from chu/bomb specifics) will check for either
- You get a special bag as the first pack also signifying you can now get chu drops from grass et al.
  - NOTE: I changed behavior since this was discussed to give whatever the original item would give instead of defaulting to 20 for the first pickup. Thought process:
  - Previous bombchus-in-logic behavior did not respect vanilla item pool. Bombchus found after the first 20 were 10 if you had less than 5 in your inventory, otherwise 5.
  - Preserving item pool values but making the first one 20 would lead to unfair racing scenarios. If player 1 finds their first pack as the 20 pack and player 2 finds their first pack as the 5 pack, player 2 still has a 20 pack somewhere in their seed that could be an advantageous refill while player 1's 5 pack will stay a 5 pack.
  - Always preserving the item pool values, even if it forces the 5 pack as the only accessible pack, keeps things fair and drops from enemies/pots/grass make up for the lack of initial chus.
- Specialized Buy Bombchus are removed from shops, on shopsanity the chu shop's 8 will be shuffled amongst the other shop pool items
- All chus remain as major items

Still undecided
- How to deal with unrestricted buyable chus, i.e. carpet guy
  - My opinion is to allow carpet guy to be the first source, but always lock bombchu bowling behind finding bombchus/buying from carpet guy. This mirrors current bombchus-in-logic behavior.
- How to deal with buyable chus that only require a bomb bag
  - I left default bombchus-in-logic behavior untouched. Shop bombchus require finding bombchus in the world first.
- Vanilla Spirit Temple 2 key bombchu route no longer applies now that bombchus can be replenished out of logic.

### CODE REVIEW

**Comment by ETR-BTF** [comment]
So let me get this straight: 
- Before getting chus/chu bags, all chu models are chu bags.
- After getting your first chu bag, all the other chu bag models revert back to regular chu models.
- Your initial chu count after obtaining a chu bag will be whatever the ammo count was for the bag you happened to pick up.
- The ammo counts of chus/chu bags are fixed upon seed creation and never change mid-seed.

Is this how it works? I'm just checking if I understand correctly.

**Comment by r0bd0g** [comment]
"Chus in logic as a setting disappears"
Being such a radical departure from vanilla behaviour I think it's important that this new chu bag behaviour would replace bombchus in logic as an option, rather than be forced on in every seed.

(And a hidden secret about using chus on things is that it can pretty annoying sometimes having to cook chus and time releases to actually blow things up. I think players might appreciate such an option for sparing them that aspect of chu usage becoming so pervasive. I think some players also enjoy having a source of explosives that aren't in logic and seeing just how far they can cheese the seed with them.)

Also, you did not update the Spirit logic correctly. The 2-keys route should stay around (and if you insist on removing bombchus in logic as a setting, you could just remove that setting from the logic, since practically speaking it's always on now). Also, with chu drops guaranteed replaceable, you can remove the entrance shuffle check. Take care also not to leave extra brackets lying around if you remove the reason for their existence. I'm not sure if 2 keys would outright obsolete the 3 keys routes for those checks as a result of this change; I haven't thought hard enough about it (since I strongly believe bombchus in logic should remain optional, in which case the 3 keys route would not be obsoleted). Your edits to the comment would also need to be changed.

**Comment by mracsys** [comment]
> So let me get this straight:
> 
> * Before getting chus/chu bags, all chu models are chu bags.
> * After getting your first chu bag, all the other chu bag models revert back to regular chu models.
> * Your initial chu count after obtaining a chu bag will be whatever the ammo count was for the bag you happened to pick up.
> * The ammo counts of chus/chu bags are fixed upon seed creation and never change mid-seed.
> 
> Is this how it works? I'm just checking if I understand correctly.

You are correct.

**Comment by cjohnson57** [comment]
> Chus in logic as a setting disappears

🥳

I agree with everything you have right now except the carpet guy which I'm still undecided on. Just doesn't feel right to get what's supposed to be a randomized item from a static source. Could you make it so you can't buy from him until you've found a bag already, maybe?

**Comment by mracsys** [comment]
> I'm not sure if 2 keys would outright obsolete the 3 keys routes for those checks as a result of this change; I haven't thought hard enough about it (since I strongly believe bombchus in logic should remain optional, in which case the 3 keys route would not be obsoleted). Your edits to the comment would also need to be changed.

If we decide to keep the 2 key route but remove the chus in logic and no entrance randomizer conditions, the logic becomes
```
...
(is_adult and has_projectile(adult) and
    ((Small_Key_Spirit_Temple, 3) or spirit_temple_shortcuts or
    (Small_Key_Spirit_Temple, 2))) or
...
```
3 key would be obsolete in that case. Are we sure we can drop the entrance rando condition? The comment discussed non-repeatable access that only exists in ER.

**Comment by mracsys** [comment]
> Could you make it so you can't buy from him until you've found a bag already, maybe?

Yep, that shouldn't be too hard.

**Comment by r0bd0g** [comment]
"# Age-unknown logic is not deadend-proof in rare cases where some non-repeatable access is used"
This comment and later are not related to bombchus in logic and shouldn't have been removed. Entrance shuffle was only checked in chus in logic to make sure that repeatable bombchus could be reached (ex: from bowling). Non-repeatable access issues still exist but those issues aren't related to chus in logic.

**Comment by mracsys** [comment]
Carpet salesman will not sell bombchus anymore until the player finds a bombchu bag from a randomized item. He is a renewable bombchu source afterwards.

If Medigoron/Carpet Salesman shuffle is on, he won't sell anything other than his randomized item (behavior unchanged from current rando). He does not require a bombchu bag for the randomized item.

**Comment by r0bd0g** [comment]
You really shouldn't need to be checking entrance shuffle there. Again, it was only there to make sure you were guaranteed to be able to access a shop where you could buy bombchus.

(Ignore the comment about entrance shuffle; it's a more general thing unrelated to the 2 keys route.)

(Without needing to check chus in logic, 2 keys would now obsolete 3 in some cases. But I think it could actually be reworded to obsolete in all cases? Would have to think about it.)

(You might also need to rework how the logic helpers for explosives work because it doesn't make sense to have has_bombchus just forward right into found_bombchus for example.)

**Comment by fenhl** [comment]
The `minor_items_as_major_chest` setting description needs to be updated since it only affects shields now. Might also make sense to rename it to “Shields in Big/Gold Chests”.

**Comment by r0bd0g** [comment]
This change really ought to be optional, which would require that option to stay as-is.
(If you're really going to force this change on everyone, you're going to need community input.)

**Comment by cjohnson57** [comment]
So I've been thinking about this and while I definitely understand the desire to make it an option, I think it's better to leave it as default.

A large reason we want this setting is because we want to remove all ambiguity of when bombchus are expected, what flags they set, and when they can be obtained (either from grass or shops) and just make it the same all the time, those things being different depending on a setting was really confusing.

**Comment by r0bd0g** [comment]
Bombchus in logic is just not fun. Game balance is thrown off by having too many replenishable sources of explosives. They're a much more cumbersome explosive to use in many cases. This PR solves only one problem with the setting, in that you always had to buy replacement chus before. If the existence of a setting is a problem, then I would say you've chosen the wrong state to keep.

File: ASM/c/get_items.c [hunk 1]
```diff
@@ -636,6 +659,19 @@ int16_t get_override_drop_id(int16_t dropId) {
         }
     }
 
+    // Chus in logic drops, convert bomb drop to bombchu drop under certain circumstances
```
Comment (cjohnson57) [inline ASM/c/get_items.c:706]: Adding this more as a reminder in myself to take the "Chus in logic" out of this comment

**Comment by cjohnson57** [comment]
Could this message be updated to specify that you need to find a chu bag first?
image

**Comment by mracsys** [comment]
Sure, but it's not trivial if I remember right from the shop actors.

**Comment by cjohnson57** [comment]
> Sure, but it's not trivial if I remember right from the shop actors.

If the message is shared between other items such as tunics, then I can see why that'd be difficult since you'd have to add a new potential message to what, iirc, is a big switch statement deciding what message to give upon selecting an item. Not a big deal either way, and always something that could be handled in a future PR if anything.

**Comment by cjohnson57** [comment]
Tested the following cases and all worked as expected:

* Bowling
    * No bag: No bowling
    * Found bag: Can bowl
    * Start with chus: Can bowl
* Shops
    * No bag, vanilla Buy chu: No buy
    * No bag, shuffled Buy chu: No buy
    * Found bag, Buy chu: Can buy
    * Start with chus, Buy chu: Can buy
    * No bag, shopsanity item chu: Can buy, turns into chu bag

**Comment by fenhl** [comment]
Can we have special deal chus appear as chu bags?

**Comment by cjohnson57** [comment]
For the same reason progressive items don't appear as their upgraded form in shops, I imagine not. A fix for that would be nice, though. 3D has them working

**Comment by fenhl** [comment]
I'm not suggesting to make the appearance adapt based on whether you already have bombchus. Just to change the appearance to a chu bag for clarity.

**Comment by cjohnson57** [comment]
We discussed on the Discord and, if possible, would like to make it so that you can do bombchu bowling with chus or bomb bag, but if you only have bomb bag then the chu reward is replaced with bombs. Some people are restless about the idea of needing to wait to do bowling until they get chus.

**Comment by r0bd0g** [comment]
I asked what happens currently with the bomb drop prize if you play with chus in and don't have a bomb bag. Does it look like a bomb drop still when you go to play the game? Does it become a blue rupee after you win it? There's probably a small rando bug there already. If you fix that, perhaps you could change the bombchu prize behave in a similar way, and maybe become a blue rupee.

(I worry the constant changing of the requirements for checks might be confusing to keep track of. I kind of just want to flip the table over and say just have the trigger be the king dodongo blue warp again. That had story justification, it wasn't just the devs saying you couldn't because they didn't want you to.)

**Comment by mracsys** [comment]
Bombchu Bowling bombchus and bomb refills now check for the corresponding bag, reverting to a purple rupee if missing. The game can be played after finding either a bomb bag or bombchus. While making the change, I also added an option to shuffle those items, off by default. The 5th item (purple rupee) is harder to shuffle because of the static prize array sizes in the two bowling actors and freestanding model draw function. The tooltip for the option states only the 3rd and 4th are shuffled to hopefully make that clear.

**Comment by cjohnson57** [comment]
Awesome work. I think all that's left is to make the bombchu bags and everything that comes with them an option, as was decided in the Discord

**Comment by mracsys** [comment]
All changes are tied to the old Bombchus in Logic setting. With it disabled, everything works how it used to work. With it enabled, it uses the changes described in this PR.

**Comment by r0bd0g** [comment]
You don't need to check for entrance shuffle off on the bombchus in logic route anymore, now that access to bombchu refills is guaranteed. (Roman fork has been able to do this for a while but has opted not to in order to keep the logic files the same between branches.)

You don't need wallet for bowling prize 4. You can win one of the replacement purple rupees to refill your money. I know that sounds arduous, but so does adding two more checks of the same thing that we already have two checks of. (Perhaps a better solution is to always present players with unclaimed prizes first, but I dunno what that might entail or exactly how that should work.)

What do the extra bowling prizes revert to after winning them?

I kind of feel like these new checks should be separate PR.

**Comment by cjohnson57** [comment]
Sorry, a couple more small changes in regard to the setting:
* I'd like to at least update the title in the GUI to be less terrible and alert people that it's changed, maybe something like "Bombchu bags (and related changes)" to make people more likely to check the tooltip to see everything it does. 
* With all the improvements this PR brings we'd like to make the new setting the default. Maybe then enough people will use it that we can finally kill bombchus in logic without too many people complaining. Probably not, but a man can dream.

> I kind of feel like these new checks should be separate PR.

After thinking on it I'm actually inclined to agree. Besides also changing something about bombchu bowling, making those two prizes into checks seems mostly unrelated to the other content of this PR

**Comment by cjohnson57** [comment]
Thanks for all the work, and sorry again for all the changes. I'll review and merge sometime this week, the weekend at the latest

**Review by cjohnson57 — APPROVED** [review]

**Pull request merged**